File: typescript/packages/core/src/core/box/read.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { BoxAccessor } from '../../accessor/box.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { PathSpec } from '../../types.ts'
import type { IndexEntry } from '../../cache/index/config.ts'
import { processBoxcanvas } from '../filetype/boxcanvas.ts'
import { processBoxnote } from '../filetype/boxnote.ts'
import { downloadFile, downloadFileStream, getExtractedText } from './api.ts'
import type { BoxTokenManager } from './_client.ts'
import { readdir } from './readdir.ts'
import { rstripSlash, stripSlash } from '../../utils/slash.ts'
import { enoent } from '../../utils/errors.ts'

const ENC = new TextEncoder()

const OFFICE_FORMAT_BY_RT: Readonly<Record<string, 'docx' | 'xlsx' | 'pptx'>> = {
  'box/gdoc': 'docx',
  'box/gsheet': 'xlsx',
  'box/gslides': 'pptx',
}

interface BoxOfficeEnvelope {
  id: string
  name: string
  format: 'docx' | 'xlsx' | 'pptx'
  size: number | null
  modified_at: string
  body_text: string
}

async function processBoxOffice(
  tm: BoxTokenManager,
  entry: IndexEntry,
  format: 'docx' | 'xlsx' | 'pptx',
): Promise<Uint8Array> {
  const bodyText = await getExtractedText(tm, entry.id)
  const envelope: BoxOfficeEnvelope = {
    id: entry.id,
    name: entry.vfsName !== '' ? entry.vfsName : entry.name,
    format,
    size: entry.size,
    modified_at: entry.remoteTime,
    body_text: bodyText,
  }
  return ENC.encode(JSON.stringify(envelope, null, 2) + '\n')
}

function eisdir(p: string): Error {
  const e = new Error(`EISDIR: ${p}`) as Error & { code: string }
  e.code = 'EISDIR'
  return e
}

export async function read(
  accessor: BoxAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<Uint8Array> {
  const prefix = path.prefix
  let p = path.original
  if (prefix !== '' && p.startsWith(prefix)) p = p.slice(prefix.length) || '/'
  const key = stripSlash(p)
  if (key === '') throw eisdir(path.original)
  if (index === undefined) throw enoent(path.original)
  const virtualKey = prefix !== '' ? `${prefix}/${key}` : `/${key}`

  let result = await index.get(virtualKey)
  if (result.entry === undefined || result.entry === null) {
    const parentKey = rstripSlash(virtualKey).replace(/\/[^/]+$/, '') || '/'
    if (parentKey !== virtualKey) {
      const parentPath = PathSpec.fromStrPath(parentKey, prefix)
      try {
        await readdir(accessor, parentPath, index)
        result = await index.get(virtualKey)
      } catch {
        // parent refresh failed; fall through to ENOENT
      }
    }
    if (result.entry === undefined || result.entry === null) throw enoent(path.original)
  }
  const rt = result.entry.resourceType
  if (rt === 'box/folder') throw eisdir(path.original)
  const officeFmt = OFFICE_FORMAT_BY_RT[rt]
  if (officeFmt !== undefined) {
    return processBoxOffice(accessor.tokenManager, result.entry, officeFmt)
  }
  const raw = await downloadFile(accessor.tokenManager, result.entry.id)
  if (rt === 'box/boxnote') return processBoxnote(raw)
  if (rt === 'box/boxcanvas') return processBoxcanvas(raw)
  return raw
}

export async function* stream(
  accessor: BoxAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): AsyncIterable<Uint8Array> {
  const prefix = path.prefix
  let p = path.original
  if (prefix !== '' && p.startsWith(prefix)) p = p.slice(prefix.length) || '/'
  const key = stripSlash(p)
  if (key === '') throw eisdir(path.original)
  if (index === undefined) throw enoent(path.original)
  const virtualKey = prefix !== '' ? `${prefix}/${key}` : `/${key}`

  let result = await index.get(virtualKey)
  if (result.entry === undefined || result.entry === null) {
    const parentKey = rstripSlash(virtualKey).replace(/\/[^/]+$/, '') || '/'
    if (parentKey !== virtualKey) {
      const parentPath = PathSpec.fromStrPath(parentKey, prefix)
      try {
        await readdir(accessor, parentPath, index)
        result = await index.get(virtualKey)
      } catch {
        // parent refresh failed; fall through to ENOENT
      }
    }
    if (result.entry === undefined || result.entry === null) throw enoent(path.original)
  }
  const rt = result.entry.resourceType
  if (rt === 'box/folder') throw eisdir(path.original)
  const officeFmt = OFFICE_FORMAT_BY_RT[rt]
  if (officeFmt !== undefined) {
    yield await processBoxOffice(accessor.tokenManager, result.entry, officeFmt)
    return
  }
  if (rt === 'box/boxnote' || rt === 'box/boxcanvas') {
    // Box-native JSON formats are tiny; fetch all then process — streaming the
    // raw JSON would force callers to JSON.parse partial bytes anyway.
    const raw = await downloadFile(accessor.tokenManager, result.entry.id)
    yield rt === 'box/boxnote' ? processBoxnote(raw) : processBoxcanvas(raw)
    return
  }
  for await (const chunk of downloadFileStream(accessor.tokenManager, result.entry.id)) {
    yield chunk
  }
}

File: python/mirage/core/gmail/read.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import json
import posixpath

from mirage.accessor.gmail import GmailAccessor
from mirage.cache.index import IndexCacheStore
from mirage.core.gmail.messages import get_attachment, get_message_processed
from mirage.core.gmail.readdir import readdir
from mirage.types import PathSpec
from mirage.utils.errors import enoent


async def read(
    accessor: GmailAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> bytes:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    virtual = path.original
    if isinstance(path, PathSpec):
        prefix = path.prefix
        path = path.original

    if prefix and path.startswith(prefix):
        rest = path[len(prefix):]
        if prefix.endswith("/") or rest == "" or rest.startswith("/"):
            path = rest or "/"
    key = path.strip("/")
    if index is None:
        raise enoent(virtual)
    virtual_key = prefix + "/" + key if prefix else "/" + key
    result = await index.get(virtual_key)
    if result.entry is None:
        parent_key = posixpath.dirname(virtual_key) or "/"
        if parent_key != virtual_key:
            parent_path = PathSpec.from_str_path(parent_key, prefix=prefix)
            try:
                await readdir(accessor, parent_path, index)
                result = await index.get(virtual_key)
            except Exception:
                pass
        if result.entry is None:
            raise enoent(virtual)
    if result.entry.resource_type in ("gmail/label", "gmail/date",
                                      "gmail/attachment_dir"):
        raise IsADirectoryError(virtual)
    if result.entry.resource_type == "gmail/attachment":
        att_dir_key = posixpath.dirname(virtual_key)
        att_dir_result = await index.get(att_dir_key)
        if att_dir_result.entry is None:
            raise enoent(virtual)
        message_id = att_dir_result.entry.id
        attachment_id = result.entry.id
        return await get_attachment(accessor.token_manager, message_id,
                                    attachment_id)
    processed = await get_message_processed(accessor.token_manager,
                                            result.entry.id)
    return json.dumps(processed, ensure_ascii=False,
                      separators=(",", ":")).encode()
